lib/ecto/query/builder/join.ex

```diff
@@ -229,6 +231,26 @@ defmodule Ecto.Query.Builder.Join do
     end
   end
 
+  defp ensure_on(nil, join) do
+    unless join[:assoc] do
+      maybe_source =
+        with {source, alias} <- join[:source],
+          source when source != nil <- source || alias do
+          " on #{inspect(source)}"
+        else
+          _ -> ""
+        end
+
+      Logger.warn(
+        "#{join[:file]}:#{join[:line]} Missing `:on` in join#{maybe_source}, defaulting to `on: true`."
+      )
```

Comment:
Please do `IO.warn` instead with passing the caller (aka __CALLER__) to retrieve the stacktrace. Using IO.warn means compilation will halt if using warnings as errors. :)

Reply:
Thanks, this is perfect, exactly what I was looking for 💜
Fixed in https://github.com/elixir-ecto/ecto/pull/4074/commits/fe18c3e28e3668de79918a6777e40e8bdd077ba0, but I needed to get the stacktrace manually to support elixir <1.14 https://github.com/elixir-ecto/ecto/pull/4074/commits/e52d35840665a1dc3aa4507ee3a83d97371bcb01